### PR TITLE
[6/6][kimi-deterministic] Use deterministic seeded coins for EAGLE rejection sampling

### PR DESCRIPTION
**Pull request by zyzshishui**
> **Stacked series (6/6, tip)**: stacked on `kimi-deterministic/5-logsoftmax-decode-logprobs` (#30821), so the diff below includes the whole series. This PR's own changes are the last 4 commits (ending 18f2f50bf — incl. the murmur_hash32 relocation to `sglang.kernels`). **CI on this PR covers the combined series.**

## Motivation

EAGLE verify draws rejection-sampling coins from `torch.rand`, so speculative outputs are not reproducible even with per-request `sampling_seed`. (Part 6/6 of the kimi-deterministic series.)

## Modifications

When `sampling_seed` is set, coins come from `murmur_hash32(seed, seq_lens, column)` (first `draft_token_num` columns: rejection coins; last column: final-sampling coin), extracted into `_seeded_verify_coins`. Unseeded requests keep `torch.rand`.

Per review: the float32 cast rounds the top 129 uint32 hashes to exactly 1.0 while the sampling kernels expect half-open `[0, 1)` coins, so the coins are clamped to the largest float32 below one (every other coin value is bit-identical, keeping verified baselines intact).

**Scope**: this seeds only the verify-side RNG. Under `speculative_use_rejection_sampling` the draft workers still select candidates via unseeded multinomial (`fast_sample` in `eagle_worker_v2`), so classic rejection-sampling mode remains non-deterministic until the draft RNG is seeded in a follow-up; top-k/greedy draft selection (e.g. EAGLE3 topk-1) is fully deterministic with this change.

## Accuracy Tests

Unseeded path unchanged. Unit tests (`test/registered/unit/spec/test_eagle_seeded_coins.py`) lock seeded-coin reproducibility, seed/position sensitivity, and the max-hash clamp. E2E: Kimi-K2 + EAGLE3 (steps 3 / topk 1 / draft 4, fa4 draft backend, 16 GPUs TP16/DP8/EP16) passes the full 7-phase invariance suite bitwise at temperature 0 and 1.0 with this series applied; non-greedy replay diverged before this change and is bitwise-stable after.

## Speed Tests and Profiling

Hash only for seeded requests; no impact otherwise.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #29979964999](https://github.com/sgl-project/sglang/actions/runs/29979964999)
Latest PR Test (Extra): :x: [Run #29979964951](https://github.com/sgl-project/sglang/actions/runs/29979964951)

### CODE REVIEW

File: python/sglang/srt/speculative/eagle_utils.py [hunk 1]
```diff
@@ -727,10 +727,32 @@ def eagle_sample(
                 "does not produce one (draft_probs missing or vocab-mismatched)."
             )
 
-        # coins for rejection sampling
-        coins = torch.rand_like(candidates, dtype=torch.float32, device=device)
-        # coins for final sampling
-        coins_for_final_sampling = torch.rand((bs,), dtype=torch.float32, device=device)
+        sampling_seed = sampling_info.sampling_seed
```
Comment (ch-wan) [inline python/sglang/srt/speculative/eagle_utils.py]: **[suggestion]** When `sampling_seed is not None`, verify-side coins become deterministic, but under `speculative_use_rejection_sampling` the draft path still selects candidates via `fast_sample` → `torch.multinomial` (see `eagle_worker_v2.py` draft_forward / draft_extend paths). Different draft tokens across runs will still produce different verify outcomes even with identical seeded coins. Tree mode without rejection sampling (deterministic `fast_topk`/`argmax` drafts) is much closer to full determinism after this change; classic rejection-sampling mode is not.

**Suggestion:** Either seed draft multinomial with the same `murmur_hash32`/`multinomial_with_seed` pattern (distinct column space from verify coins), or explicitly document in the PR/series that part 6 only closes the verify-coin RNG and rejection-sampling drafts remain non-deterministic until a follow-up.

File: python/sglang/srt/speculative/eagle_utils.py [hunk 2]
```diff
@@ -727,10 +727,32 @@ def eagle_sample(
...
+        if sampling_seed is not None:
+            from sglang.srt.layers.utils.hash import murmur_hash32
+
+            cols = torch.arange(
+                verify_input.draft_token_num + 1, device=device, dtype=torch.int64
+            )
+            hashed = murmur_hash32(
```
Comment (ch-wan) [inline python/sglang/srt/speculative/eagle_utils.py]: **[suggestion]** No unit or e2e test asserts that fixed `sampling_seed` yields identical EAGLE accept lengths / bonus tokens across runs (or that unseeded still uses `torch.rand`). The PR body acknowledges this. A small GPU unit test that mocks `target_probs`/`draft_probs`/`candidates` and checks `eagle_sample` coin-driven outputs for two identical seeds would catch regressions in column split, salt choice, and dtype conversion without a full server.

**Suggestion:** Add a focused test (kernel-level or `eagle_sample` with fixed logits) for (seeded invariance) and (seeded vs unseeded branch). Prefer also an e2e speculative invariance run once the series is complete.

File: python/sglang/srt/speculative/eagle_utils.py [hunk 3]
```diff
@@ -727,10 +727,32 @@ def eagle_sample(
...
+                sampling_seed.to(torch.uint64), batch.seq_lens.to(torch.uint64), cols
```
Comment (ch-wan) [inline python/sglang/srt/speculative/eagle_utils.py]: **[bug]** `hashed / UINT32_MAX` is not guaranteed to remain in `[0, 1)` after conversion to float32. The highest 129 uint32 hash values round to exactly `1.0` (about one coin in 33 million). Both sampling kernels expect a half-open uniform range. In the rejection-sampling final CDF, `coin == 1.0` makes the target equal the complete probability sum; the strict `cumsum > target` comparison finds no token and falls back to `VOCAB_SIZE - 1`, potentially selecting a zero-probability token. Endpoint rejection coins can also bias draft acceptance.

**Suggestion:** Clamp the float32 coins to the largest representable value below one, or construct them from truncated hash bits so the range is inherently `[0, 1)`. Add a regression test covering the maximum hash value.

**Review by ch-wan — APPROVED** [review]
LGTM
